File: native-modules/react-native-range-downloader/android/src/main/java/com/margelo/nitro/reactnativerangedownloader/FirmwareArtifactRegistry.kt
package com.margelo.nitro.reactnativerangedownloader

import okhttp3.Call
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ExecutorService
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong
import java.util.concurrent.atomic.AtomicReference

internal enum class FirmwareArtifactRegistryState {
  NOT_FOUND,
  QUEUED,
  DOWNLOADING,
  VERIFYING,
  COMPLETED,
  CANCELLED,
  FAILED,
}

internal data class FirmwareArtifactRegistrySnapshot(
  val state: FirmwareArtifactRegistryState,
  val downloadedBytes: Long,
  val expectedSize: Long,
  val artifact: StoredArtifactMetadata? = null,
  val errorCode: String? = null,
  val errorMessage: String? = null,
  val retryable: Boolean? = null,
)

internal class FirmwareArtifactCancellation {
  val cancelled = AtomicBoolean(false)
  private val calls = ConcurrentHashMap.newKeySet<Call>()

  @Volatile
  private var executor: ExecutorService? = null

  fun attach(call: Call) {
    calls += call
    if (cancelled.get()) {
      call.cancel()
    }
  }

  fun detach(call: Call) {
    calls -= call
  }

  fun attach(executor: ExecutorService) {
    this.executor = executor
    if (cancelled.get()) {
      executor.shutdownNow()
    }
  }

  fun cancelAndWait() {
    cancelled.set(true)
    calls.forEach(Call::cancel)
    executor?.let { pool ->
      pool.shutdownNow()
      try {
        pool.awaitTermination(5, TimeUnit.SECONDS)
      } catch (_: InterruptedException) {
        Thread.currentThread().interrupt()
      }
    }
  }
}

internal class FirmwareArtifactRegistry {
  private data class TaskKey(
    val leaseRef: String,
    val taskId: String,
  )

  private class Entry(
    val fingerprint: String,
    val expectedSize: Long,
  ) {
    val cancellation = FirmwareArtifactCancellation()
    val future = CompletableFuture<StoredArtifactMetadata>()
    val downloadedBytes = AtomicLong(0)
    val state = AtomicReference(FirmwareArtifactRegistryState.QUEUED)
    val errorCode = AtomicReference<String?>(null)
    val errorMessage = AtomicReference<String?>(null)
    val retryable = AtomicReference<Boolean?>(null)
  }

  private val entries = ConcurrentHashMap<TaskKey, Entry>()
  private val activeArtifacts = ConcurrentHashMap<String, TaskKey>()

  fun runOrAttach(
    leaseRef: String,
    taskId: String,
    artifactId: String,
    fingerprint: String,
    expectedSize: Long,
    operation: (
      cancellation: FirmwareArtifactCancellation,
      update: (FirmwareArtifactRegistryState, Long) -> Unit,
    ) -> StoredArtifactMetadata,
  ): StoredArtifactMetadata {
    val key = TaskKey(leaseRef, taskId)
    val mine = Entry(fingerprint, expectedSize)
    var incumbent = entries.putIfAbsent(key, mine)
    if (incumbent?.future?.isCompletedExceptionally == true &&
      entries.replace(key, incumbent, mine)
    ) {
      incumbent = null
    }
    val entry = incumbent ?: mine
    if (entry.fingerprint != fingerprint) {
      throw FirmwareArtifactStoreException.invalidMetadata()
    }
    if (incumbent != null) {
      return entry.future.get()
    }

    val activeOwner = activeArtifacts.putIfAbsent(artifactId, key)
    if (activeOwner != null && activeOwner != key) {
      val error = FirmwareArtifactTransferException(
        code = "ARTIFACT_BUSY",
        retryable = true,
        discardStaging = false,
        message = "Another firmware task owns this artifact",
      )
      fail(entry, error)
      throw error
    }
    try {
      val result = operation(entry.cancellation) { state, downloadedBytes ->
        entry.downloadedBytes.set(downloadedBytes.coerceAtLeast(0))
        entry.state.set(state)
      }
      entry.downloadedBytes.set(result.actualSize ?: expectedSize)
      entry.state.set(FirmwareArtifactRegistryState.COMPLETED)
      entry.future.complete(result)
      pruneCompletedEntries()
      return result
    } catch (error: Exception) {
      fail(entry, error)
      throw error
    } finally {
      activeArtifacts.remove(artifactId, key)
    }
  }

  fun status(leaseRef: String, taskId: String): FirmwareArtifactRegistrySnapshot? {
    val entry = entries[TaskKey(leaseRef, taskId)] ?: return null
    return FirmwareArtifactRegistrySnapshot(
      state = entry.state.get(),
      downloadedBytes = entry.downloadedBytes.get(),
      expectedSize = entry.expectedSize,
      artifact = if (entry.future.isDone && !entry.future.isCompletedExceptionally) {
        entry.future.getNow(null)
      } else {
        null
      },
      errorCode = entry.errorCode.get(),
      errorMessage = entry.errorMessage.get(),
      retryable = entry.retryable.get(),
    )
  }

  fun cancel(leaseRef: String, taskId: String) {
    entries[TaskKey(leaseRef, taskId)]?.cancellation?.cancelAndWait()
  }

  private fun fail(entry: Entry, error: Exception) {
    val transferError = error as? FirmwareArtifactTransferException
    val storeError = error as? FirmwareArtifactStoreException
    entry.errorCode.set(
      transferError?.code
        ?: storeError?.code
        ?: "ARTIFACT_DOWNLOAD_FAILED"
    )
    entry.errorMessage.set(error.message)
    entry.retryable.set(transferError?.retryable ?: storeError?.retryable)
    entry.state.set(
      if (entry.cancellation.cancelled.get()) {
        FirmwareArtifactRegistryState.CANCELLED
      } else {
        FirmwareArtifactRegistryState.FAILED
      }
    )
    entry.future.completeExceptionally(error)
    pruneCompletedEntries()
  }

  private fun pruneCompletedEntries() {
    if (entries.size <= MAX_RETAINED_TASKS) return
    entries.entries
      .asSequence()
      .filter { it.value.future.isDone }
      .take(entries.size - MAX_RETAINED_TASKS)
      .forEach { entries.remove(it.key, it.value) }
  }

  companion object {
    private const val MAX_RETAINED_TASKS = 128
    val processInstance = FirmwareArtifactRegistry()
  }
}

File: native-modules/react-native-range-downloader/ios/FirmwareArtifactRegistry.swift
import Foundation

enum FirmwareArtifactRegistryState {
  case notFound
  case queued
  case downloading
  case verifying
  case completed
  case cancelled
  case failed
}

struct FirmwareArtifactRegistrySnapshot {
  let state: FirmwareArtifactRegistryState
  let downloadedBytes: Int64
  let expectedSize: Int64
  let artifact: StoredFirmwareArtifactMetadata?
  let errorCode: String?
  let errorMessage: String?
  let retryable: Bool?
}

final class FirmwareArtifactRegistry {
  static let shared = FirmwareArtifactRegistry()

  private struct TaskKey: Hashable {
    let leaseRef: String
    let taskId: String
  }

  private final class Entry {
    let fingerprint: String
    let expectedSize: Int64
    var task: Task<StoredFirmwareArtifactMetadata, Error>?
    var state: FirmwareArtifactRegistryState = .queued
    var downloadedBytes: Int64 = 0
    var artifact: StoredFirmwareArtifactMetadata?
    var errorCode: String?
    var errorMessage: String?
    var retryable: Bool?

    init(fingerprint: String, expectedSize: Int64) {
      self.fingerprint = fingerprint
      self.expectedSize = expectedSize
    }
  }

  private let lock = NSLock()
  private var entries: [TaskKey: Entry] = [:]
  private let activeArtifacts = FirmwareArtifactRunOwnership<TaskKey>()

  init() {}

  func runOrAttach(
    leaseRef: String,
    taskId: String,
    artifactId: String,
    fingerprint: String,
    expectedSize: Int64,
    operation: @escaping (
      _ update: @escaping (FirmwareArtifactRegistryState, Int64) -> Void
    ) async throws -> StoredFirmwareArtifactMetadata
  ) async throws -> StoredFirmwareArtifactMetadata {
    let key = TaskKey(leaseRef: leaseRef, taskId: taskId)
    let selection: (
      entry: Entry,
      task: Task<StoredFirmwareArtifactMetadata, Error>,
      ownsTask: Bool
    ) = try lock.withFirmwareLock {
      if let existing = entries[key],
         existing.fingerprint == fingerprint,
         existing.state != .failed,
         existing.state != .cancelled,
         let existingTask = existing.task {
        return (existing, existingTask, false)
      }
      if let existing = entries[key], existing.fingerprint != fingerprint {
        throw FirmwareArtifactStoreError.invalidMetadata
      }
      guard activeArtifacts.acquire(artifactId: artifactId, owner: key) else {
        throw FirmwareArtifactTransferError.busy
      }
      let entry = Entry(fingerprint: fingerprint, expectedSize: expectedSize)
      let createdTask = Task { [activeArtifacts = self.activeArtifacts] in
        defer {
          activeArtifacts.release(artifactId: artifactId, owner: key)
        }
        return try await operation { [weak self, weak entry] state, downloadedBytes in
          guard let self, let entry else { return }
          self.lock.withFirmwareLock {
            entry.state = state
            entry.downloadedBytes = max(0, downloadedBytes)
          }
        }
      }
      entry.task = createdTask
      entries[key] = entry
      return (entry, createdTask, true)
    }
    let entry = selection.entry
    let task = selection.task
    let ownsTask = selection.ownsTask

    do {
      let result = try await task.value
      if ownsTask {
        lock.withFirmwareLock {
          entry.state = .completed
          entry.downloadedBytes = result.actualSize ?? expectedSize
          entry.artifact = result
          pruneCompletedEntries()
        }
      }
      return result
    } catch {
      if ownsTask {
        let transferError = error as? FirmwareArtifactTransferError
        let storeError = error as? FirmwareArtifactStoreError
        lock.withFirmwareLock {
          entry.state = task.isCancelled ? .cancelled : .failed
          entry.errorCode = transferError?.code ??
            storeError?.code ??
            "ARTIFACT_DOWNLOAD_FAILED"
          entry.errorMessage = error.localizedDescription
          entry.retryable = transferError?.retryable ?? storeError?.retryable
          pruneCompletedEntries()
        }
      }
      throw error
    }
  }

  func update(
    leaseRef: String,
    taskId: String,
    state: FirmwareArtifactRegistryState,
    downloadedBytes: Int64
  ) {
    lock.withFirmwareLock {
      guard let entry = entries[TaskKey(leaseRef: leaseRef, taskId: taskId)] else {
        return
      }
      entry.state = state
      entry.downloadedBytes = max(0, downloadedBytes)
    }
  }

  func fail(
    leaseRef: String,
    taskId: String,
    error: FirmwareArtifactTransferError
  ) {
    lock.withFirmwareLock {
      guard let entry = entries[TaskKey(leaseRef: leaseRef, taskId: taskId)] else {
        return
      }
      entry.state = .failed
      entry.errorCode = error.code
      entry.errorMessage = error.localizedDescription
      entry.retryable = error.retryable
    }
  }

  func status(
    leaseRef: String,
    taskId: String
  ) -> FirmwareArtifactRegistrySnapshot? {
    return lock.withFirmwareLock {
      guard let entry = entries[TaskKey(leaseRef: leaseRef, taskId: taskId)] else {
        return nil
      }
      return FirmwareArtifactRegistrySnapshot(
        state: entry.state,
        downloadedBytes: entry.downloadedBytes,
        expectedSize: entry.expectedSize,
        artifact: entry.artifact,
        errorCode: entry.errorCode,
        errorMessage: entry.errorMessage,
        retryable: entry.retryable
      )
    }
  }

  func cancel(leaseRef: String, taskId: String) {
    lock.withFirmwareLock {
      let entry = entries[TaskKey(leaseRef: leaseRef, taskId: taskId)]
      entry?.state = .cancelled
      entry?.task?.cancel()
    }
  }

  private func pruneCompletedEntries() {
    guard entries.count > 128 else { return }
    let removable = entries.filter {
      $0.value.state == .completed ||
        $0.value.state == .failed ||
        $0.value.state == .cancelled
    }
    for (key, _) in removable.prefix(entries.count - 128) {
      entries.removeValue(forKey: key)
    }
  }
}
